Split command queue into issue/completion regions and update EnqueueReadBuffer to read from completion queue

abhullar-tt opened the pull request.

abhullar-tt committed:
#3943: Reduce issue region of cq to 75% of total cq size

abhullar-tt committed:
#3943: Update CQ consumer kernel to write data requested into completion queue and enable support for reads that are larger than available space in completion region

abhullar-tt commented on tt_metal/impl/dispatch/device_command.hpp:
```diff
@@ -15,6 +15,11 @@ class DeviceCommand {
     static constexpr uint32_t HUGE_PAGE_SIZE = 1024 * 1024 * 1024;
+
+    // COMMAND_ISSUE_REGION_SIZE and COMMAND_COMPLETION_REGION_SIZE should be configurable with some default split
+    static constexpr uint32_t COMMAND_ISSUE_REGION_SIZE = 805306368;
```
if we want it to be configurable should we allow completion size to go to 0? and in that case EnqueueReadBuffer populates the issue region. Or should there be some min size of completion region.

abhullar-tt replied:
I prefer one data path (to device via issue and from device via completion) but a reduced issue region could lead to more stalls

DrJessop replied:
I think this shouldn't be a constexpr value. Rather, I think there should be some defaults when constructing the command queue, and someone those defaults propagate to send kernel to device where we pass it as a compile time arg to the producer kernel.

TT-billteng replied:
how was this number calculated?

abhullar-tt replied:
atm just arbitrarily assigned 75% of the CQ to issue region because I figured we would be doing less read backs from device

TT-billteng replied:
can we change it to `0.75 * HUGE_PAGE_SIZE` and explain psuedo-reason behind 75%

abhullar-tt replied:
before merging this I am planning on removing the constexpr vars representing completion/issue sizes. Instead these will be optionally user configurable. The split can be passed into `GetCommandQueue` (from https://github.com/tenstorrent-metal/tt-metal/pull/4077) 

and I will add a comment to explain the default split

DrJessop commented on tt_metal/hw/inc/dataflow_api.h:
```diff
@@ -1525,11 +1528,50 @@ void cq_pop_front(uint32_t cmd_size_B) {
+    } while (
+        cq_write_interface.completion_fifo_wr_ptr < rd_ptr and cq_write_interface.completion_fifo_wr_ptr + data_size_16B > rd_ptr or
+        (rd_toggle != cq_write_interface.completion_fifo_wr_toggle and cq_write_interface.completion_fifo_wr_ptr == rd_ptr)
+    );
```
we should make it explicit that the rd_ptr is the completion_rd_ptr

DrJessop commented on tt_metal/hw/inc/dataflow_api.h:
```diff
@@ -1525,11 +1528,50 @@ void cq_pop_front(uint32_t cmd_size_B) {
+    uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    cq_write_interface.completion_fifo_wr_ptr += push_size_16B;
+    if (cq_write_interface.completion_fifo_wr_ptr >= cq_write_interface.completion_fifo_limit) {
+        cq_write_interface.completion_fifo_wr_ptr = DeviceCommand::COMMAND_ISSUE_REGION_SIZE >> 4;
```
DeviceCommand::COMMAND_ISSUE_REGION_SIZE should be passed in as a compile-time variable.

DrJessop commented on tt_metal/hw/inc/dataflow_api.h:
```diff
@@ -1525,11 +1528,50 @@ void cq_pop_front(uint32_t cmd_size_B) {
+    volatile tt_l1_ptr uint32_t* wr_ptr_addr = get_cq_completion_write_ptr();
+    wr_ptr_addr[0] = wr_ptr_and_toggle;
+    noc_async_write(CQ_COMPLETION_WRITE_PTR, pcie_address, 4);
+    noc_async_write_barrier();
```
I added a write flush barrier that can be used here.

abhullar-tt replied:
do you mean to use `noc_async_writes_flushed()` instead of `noc_async_write_barrier()`?

DrJessop replied:
Yes, the noc_async_writes_flushed API just ensures that the data being written has left source, but no guarantees that it has arrived to the destination.

abhullar-tt replied:
don't we want to ensure that it arrived at the dst?

DrJessop replied:
We can keep the barrier for now and revisit.

DrJessop commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -860,33 +859,68 @@ void convert_interleaved_to_sharded_on_host(const void * host, const int num_pag
+    uint32_t unpadded_bytes_to_read = buffer.size();
+    uint32_t unpadded_dst_offset = 0;
+
+    while (total_bytes_to_read > 0) {
```
should be in terms of num pages

DrJessop reviewed (changes requested):
Adding initial set of comments just so you can see them. Will continue with a second pass tomorrow.

abhullar-tt commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:
```diff
@@ -318,8 +318,14 @@ TEST_F(CommandQueueFixture, TestPageSizeTooLarge) {
+    uint32_t max_command_size = DeviceCommand::COMMAND_ISSUE_REGION_SIZE - CQ_START;
+    uint32_t buffer = 14240;
+    uint32_t buffer_size = max_command_size - (buffer + DeviceCommand::NUM_BYTES_IN_DEVICE_COMMAND);
+    TT_ASSERT(buffer_size % page_size == 0);
```
yeah good point

abhullar-tt committed:
#3943: Issue multiple enqueue write commands when buffer to write does not fit in current available space in issue region

abhullar-tt committed:
#3943: Pass CQ issue and completion sizes as compile time args to producer/consumer kernels

abhullar-tt committed:
#3943: update wrap command to accept wrap region to allow host to
request completion queue write ptr wrap

wrapping issue uses full size of issue queue when wrapping on issue queue to ensure the writer pointer gets reset to head of issue q

abhullar-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -899,6 +912,12 @@ void CommandQueue::enqueue_read_buffer(Buffer& buffer, void* dst, bool blocking)
         uint32_t available_space_bytes = DeviceCommand::HUGE_PAGE_SIZE - (get_cq_completion_wr_ptr(this->device->id()) << 4);
+        if (available_space_bytes < padded_page_size) {
+            // wrap the completion region because a single page won't fit in available space
+            this->wrap(DeviceCommand::WrapRegion::COMPLETION);
```
@DrJessop in this commit I resolve a hang that I was hitting because space left in completion region was less than a page. In that case host issues a wrap to completion region. 

I updated wrap index to accept one of: 
```
    enum class WrapRegion : uint8_t {
        NONE = 0,
        ISSUE = 1,
        COMPLETION = 2
    };
// defined in device_command.hpp
```

DrJessop replied:
Shouldn't this be a function of the number of pages that you're trying to read in one shot into the completion queue (rather than just 1 padded page size)?

abhullar-tt replied:
ah yeah so i was using the available size to figure out how big of a read to issue but I can update the logic to try and read in one shot and only issue multiple reads if buffer size is larger than completion q size

abhullar-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -727,17 +727,30 @@ const DeviceCommand EnqueueWrapCommand::assemble_device_command(uint32_t) {
 
-    // Since all of the values will be 0, this will be equivalent to
-    // a bunch of NOPs
-    vector<uint32_t> command_vector(space_left / sizeof(uint32_t), 0);
```
@DrJessop why can't we send the wrap command similar to the case for wrapping the completion? 

BTW I would like to consolidate these two paths but just pushed this to fix a hang I was hitting

abhullar-tt commented on tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:
```diff
@@ -104,23 +104,30 @@ void cq_reserve_back(uint32_t data_size_B) {
+FORCE_INLINE
+void cq_push_back(const uint32_t command_issue_region_size, uint32_t push_size_B, uint32_t min_space_required_B) {
     uint32_t push_size_16B = align(push_size_B, 32) >> 4;
+    uint32_t min_space_required_16B = align(min_space_required_B, 32) >> 4;
```
Im not sure if I should do this but there can be a case where we split our buffer read into multiple enqueue_read_buffer commands but during that process we could update our write ptr such that there isn't enough space for one page. In that case we can wrap the completion queue earlier than the next host iteration of computing how much available space there is by seeing if updated wr ptr would leave less than `min_space_required_B` in the completion queue. The completion queue read ptr is updated in same manner

Issue with this is that if the current read is the last read for the buffer and the next buffer to read has smaller page size than current, then we may have unnecessarily wrapped

If I remove this logic then host will just see there is not enough space and issue a wrap to completion (might be more straightforward to follow). Let me know your thoughts @DrJessop

DrJessop replied:
I'm not following. If we issue multiple enqueue read buffer commands, why can't we just have a loop that has the following:

```c++
while (not done) {
 if (wrap_condition)
  wrap();
 EnqueueReadBuffer command(...); 
 command.process()
}
```

abhullar-tt replied:
yes will update logic

abhullar-tt committed:
#3943: Make issue q and completion q sizes configurable by passing them into CreateDevice

abhullar-tt committed:
#3943: Make wrap on completion queue blocking so that host sees the updated completion queue write ptr to compute num pages that can be read

DrJessop commented on tt_metal/tt_metal.cpp:
```diff
@@ -562,11 +562,16 @@ size_t GetNumPCIeDevices() {
     const char *TT_METAL_SLOW_DISPATCH_MODE = std::getenv("TT_METAL_SLOW_DISPATCH_MODE");
     if (TT_METAL_SLOW_DISPATCH_MODE == nullptr) {
-        detail::GLOBAL_CQ = std::make_unique<CommandQueue>(dev);
+        if (command_issue_queue_split.has_value() and command_issue_queue_split.value() <= 0) {
```
Should check that the value is between 0 and 1.

DrJessop commented on tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:
```diff
@@ -90,24 +146,32 @@ FORCE_INLINE void write_buffers(
+
+        if (buffer_type == BufferType::SYSTEM_MEMORY) {
+            cq_reserve_back(num_pages * page_size);
+        }
```
potentially we should rename this buffer type to be "completion buffer"

abhullar-tt replied:
makes sense because a SYSTEM_MEMORY buffer will always write into the completion queue

abhullar-tt replied:
won't make this change because enqueue write buffer src buff type is SYSTEM_MEMORY

DrJessop commented on tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:
```diff
@@ -116,7 +180,10 @@ FORCE_INLINE void write_buffers(
+            page_id += num_to_write;
+        }
+        if (buffer_type == BufferType::SYSTEM_MEMORY) {
+            cq_push_back(command_issue_region_size, num_pages * page_size);
```
I think it would be more readable with "completion_queue_push_back"

abhullar-tt replied:
makes sense. I will rename all the `cq_*` APIs on device and host to be `issue_queue_*` and `completion_queue_*`

DrJessop commented on tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:
```diff
@@ -73,8 +76,61 @@ uint32_t get_read_ptr(bool db_buf_switch) {
+
+    // Notify host of updated completion wr ptr
+    notify_host_of_cq_completion_write_pointer();
+}
```
Potentially we should now indicate that this is a "completion queue API"

abhullar-tt replied:
`notify_host_of_cq_completion_write_pointer` -> `notify_host_of_completion_queue_write_pointer`?

DrJessop commented on tt_metal/impl/dispatch/kernels/command_queue_producer.hpp:
```diff
@@ -6,6 +6,52 @@
+
+    notify_host_of_cq_issue_read_pointer();
+}
+
```
these would be more readable with "issue_queue_wait/pop_front"

DrJessop reviewed (changes requested):
First set of changes to look over. Will maybe need another review.

abhullar-tt committed:
#3943: Rename cq_* apis to indicate whether they interact with issue queue or completion queue

abhullar-tt commented on tt_metal/host_api.hpp:
```diff
@@ -56,11 +56,12 @@ size_t GetNumPCIeDevices();
+ * | command_issue_queue_split | Percentage of the command queue that is dedicated for issuing commands. Issue queue size is rounded to be 32B aligned.<br><br>Remaining space is dedicated for completion queue.<br><br>Commands and data being sent to device are pushed to issue queue, data requested from device is pushed to completion queue.<br><br>Smaller issue queues can lead to more stalls due to ...
  * */
-Device *CreateDevice(chip_id_t device_id, const std::vector<uint32_t>& l1_bank_remap = {});
+Device *CreateDevice(chip_id_t device_id, std::optional<float> command_issue_queue_split = std::nullopt, const std::vector<uint32_t>& l1_bank_remap = {});
```
@TT-billteng I added the split as config to CreateDevice because in my "CQ per device" PR #4077 , the dispatch kernels are sent on device init and the issue/completion queue sizes are sent as compile time args to the dispatch kernels

TT-billteng replied:
who would be the right audience that would want to change this param? I feel like this is too much detail in an API as often used as `CreateDevice`

DrJessop commented on tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:
```diff
@@ -318,12 +318,98 @@ TEST_F(CommandQueueFixture, TestPageSizeTooLarge) {
+    EXPECT_EQ(src_2, result_2);
+}
+
+
```
We should probably have tests for the extremes (where the completion queue is sized to 0% and issue queue sized to 100%)

DrJessop commented on tt_metal/hostdevcommon/common_runtime_address_map.h:
```diff
@@ -90,38 +90,38 @@ constexpr static std::uint32_t OP_INFO_SIZE      = 280; // So far, holds up to 1
 constexpr static std::uint64_t DISPATCH_MESSAGE_REMOTE_SENDER_ADDR = 110848;
 
-constexpr static std::uint32_t COMMAND_PTR_SHARD_IDX = 6;
+constexpr static std::uint32_t COMMAND_PTR_SHARD_IDX = 8;
```
was this a bug?

abhullar-tt replied:
No, I added src and dst page offsets to support multiple read/write commands for one buffer

abhullar-tt committed:
#3943: don't make issue/completion configurable

abhullar-tt committed:
#3943: fixes to completion queue changes after rebase

TT-billteng reviewed (approved):
Looks great :)
